src/webgpu/shader/execution/shader_io/compute_builtins.spec.ts

```diff
@@ -254,3 +255,424 @@ g.test('inputs')
+ * @param max The maximum subgroup size allowed
+ * @param invocations The number of invocations in a workgroup
+ */
+function checkSubgroupSizeConsistency(
+  data: Uint32Array,
+  compare: Uint32Array,
+  min: number,
+  max: number,
+  invocations: number
+): Error | undefined {
+  const subgroupSize = data[0];
+  if (popcount(subgroupSize) !== 1) {
+    return new Error(`Subgroup size '${subgroupSize}' is not a power of two`);
+  }
+  if (subgroupSize < min) {
+    return new Error(`Subgroup size '${subgroupSize}' is less than minimum '${min}'`);
+  }
+  if (max < subgroupSize) {
+    return new Error(`Subgroup size '${subgroupSize}' is greater than maximum '${max}'`);
+  }
+
+  // Check that remaining invocations record a consistent subgroup size.
+  for (let i = 1; i < data.length; i++) {
+    if (data[i] !== subgroupSize) {
+      return new Error(
+        ErrorMsg(`Invocation ${i}: subgroup size inconsistency`, data[i], subgroupSize)
+      );
+    }
+  }
+
+  // Assumes workgroups are divided such that there is a single partial subgroup if any.
+  const partialSize = invocations % subgroupSize;
+  let fullSize = 0;
+  if (subgroupSize <= invocations) {
+    fullSize = Math.floor(invocations / subgroupSize) * subgroupSize;
+  }
+  let fullCount = 0;
+  let partialCount = 0;
+  for (let i = 0; i < compare.length; i++) {
+    if (i % invocations === 0) {
+      if (i !== 0) {
+        // Check intermediate workgroup counts.
+        if (fullCount !== fullSize) {
+          return new Error(ErrorMsg(`Unexpected number of full invocations`, fullCount, fullSize));
+        }
+        if (partialCount !== partialSize) {
+          return new Error(
+            ErrorMsg(`Unexpected number of partial invocations`, partialCount, partialSize)
+          );
+        }
+      }
+
+      // Reset for new workgroup.
+      fullCount = 0;
+      partialCount = 0;
+    }
+
+    if (compare[i] === subgroupSize) {
+      fullCount++;
+    } else {
+      partialCount++;
+
+      // Check that all partial invocations are consistent.
+      if (compare[i] !== partialSize) {
+        return new Error(ErrorMsg(`Partial subgroup size incorrect`, compare[i], partialSize));
+      }
+    }
+  }
+
+  // Check final workgroup counts.
+  if (fullCount !== fullSize) {
+    return new Error(ErrorMsg(`Unexpected number of full invocations`, fullCount, fullSize));
+  }
+  if (partialCount !== partialSize) {
+    return new Error(
+      ErrorMsg(`Unexpected number of partial invocations`, partialCount, partialSize)
+    );
+  }
+
+  return undefined;
+}
+
+const kWGSizes = [
+  [1, 1, 1],
+  [4, 1, 1],
+  [8, 1, 1],
+  [16, 1, 1],
+  [32, 1, 1],
+  [64, 1, 1],
+  [128, 1, 1],
+  [256, 1, 1],
+  [1, 4, 1],
+  [1, 8, 1],
+  [1, 16, 1],
+  [1, 32, 1],
+  [1, 64, 1],
+  [1, 128, 1],
+  [1, 256, 1],
+  [1, 1, 4],
+  [1, 1, 8],
+  [1, 1, 16],
+  [1, 1, 32],
+  [1, 1, 64],
+  [3, 3, 3],
+  [4, 4, 4],
+  [16, 16, 1],
+  [16, 1, 16],
+  [1, 16, 16],
+  [15, 3, 3],
```

Comment:
These odd ones at the end should reveal the  repeated-ragged shape issues. Good.